Mathlib/Algebra/Category/ModuleCat/Presheaf.lean

```diff
@@ -203,4 +234,194 @@ instance : Faithful (toPresheaf R) where
     simp only [toPresheaf_obj, toPresheaf_map_app] at eq
     simp only [← toAddMonoidHom_coe, eq]
 
+variable (R)
+
+/-- Evaluation on an object `X` gives a functor
+`PresheafOfModules R ⥤ ModuleCat (R.obj X)`. -/
+@[simps]
+def evaluation (X : Cᵒᵖ) : PresheafOfModules.{v} R ⥤ ModuleCat (R.obj X) where
+  obj M := M.obj X
+  map f := f.app X
+
+variable {R}
+
+/-- Given a presheaf of modules `M` on a category `C` and `f : X ⟶ Y` in `Cᵒᵖ`, this
+is the restriction map `M.obj X ⟶ M.obj Y`, considered as a linear map to
+the restriction of scalars of `M.obj Y`. -/
+noncomputable def restrictionApp {X Y : Cᵒᵖ} (f : X ⟶ Y) (M : PresheafOfModules.{v} R) :
+    M.obj X ⟶ (ModuleCat.restrictScalars (R.map f)).obj (M.obj Y) :=
+  ModuleCat.semilinearMapAddEquiv (R.map f) _ _ (M.map f)
+
+lemma restrictionApp_apply {X Y : Cᵒᵖ} (f : X ⟶ Y) (M : PresheafOfModules R) (x : M.obj X) :
+    restrictionApp f M x = M.map f x := by
+  rfl
+
+variable (R)
+
+/-- The restriction natural transformation on presheaves of modules, considered as linear maps
+to restriction of scalars. -/
+@[simps]
+noncomputable def restriction {X Y : Cᵒᵖ} (f : X ⟶ Y) :
+    evaluation R X ⟶ evaluation R Y ⋙ ModuleCat.restrictScalars (R.map f) where
+  app := restrictionApp f
+  naturality := fun M N φ => by
+    ext x
+    exact (congr_hom (φ.hom.naturality f) x).symm
+
+variable {R}
+
+attribute [local simp] restrictionApp_apply
+
+lemma restriction_app_id (M : PresheafOfModules R) (X : Cᵒᵖ) :
+    restrictionApp (𝟙 X) M =
+      (ModuleCat.restrictScalarsId' (R.map (𝟙 X)) (R.map_id X)).inv.app (M.obj X) := by aesop
+
+lemma restriction_app_comp (M : PresheafOfModules R) {X Y Z : Cᵒᵖ} (f : X ⟶ Y) (g : Y ⟶ Z) :
+    restrictionApp (f ≫ g) M =
+      restrictionApp f M ≫
+        (ModuleCat.restrictScalars (R.map f)).map (restrictionApp g M) ≫
+        (ModuleCat.restrictScalarsComp' _ _ _ (R.map_comp f g)).inv.app (M.obj Z) := by aesop
+
+end PresheafOfModules
+
+variable (R) in
+/-- This structure contains the data and axioms in order to
+produce a `PresheafOfModules R` from a collection of types
+equipped with module structures over the various rings `R.obj X`.
+(See the constructor `PresheafOfModules.mk'`.) -/
+structure CorePresheafOfModules where
+  /-- the datum of a type for each object in `Cᵒᵖ` -/
+  obj (X : Cᵒᵖ) : Type v
```

Comment:
I am wondering whether it should be `X : Cᵒᵖ` or `X : C`? I guess, the first is more natural for building up the theory while the former might be more convenient to use?

Reply:
Changing `Cᵒᵖ` to `C` may only create more issues while developing the theory, so that I would not rather not experiment with this. It seems to me that the possible gain for the future applications is very close to zero.